### PR TITLE
Builder flow for Deneb & Blobs

### PR DESCRIPTION
**Pull request by jimmygchen**
## Issue Addressed

Addresses #3689

Builder spec: https://github.com/ethereum/builder-specs/pull/61
4844 Builder flow: https://hackmd.io/@jimmygchen/B1dLR74Io

## Proposed Changes

- [x] Add new `Sidecar` and `BlockProposal` traits that allows parameterizing sidecar types in `BlockContents`, `SignedBlockContents` and to reduce code duplication required for blinded blob flow.

### Beacon Node

- `produceBlindedBlock` endpoint updates
  - [x] Handle new Deneb `BuilderBid` response from builder endpoint (new `BlindedBlobsBundle` type)
  - [x] Build `BlockContents` response (containing `kzg_commitments`, `proof` and `blinded_blob_sidecars`)
- `submitBlindedBlock endpoint` updates
  - [x] Handle `SignedBlockContents` request and forward to builder
  - [x] Handle unblinded `ExecutionPayloadAndBlobsBundle` response from builder endpoint, and publish block
- [x] Caching blobs from local EE in the blinded flow
 
### Validator Client

- [x] Handle `produceBlindedBlock` response from Beacon API
- [x] Individual signing of `blinded_blob_sidecars`
- [x] Build `SignedBlockContents` and submit to Beacon API (`submitBlindedBlock`)

### CODE REVIEW

**Comment by jimmygchen** [comment]
Hey @realbigsean @pawanjay176 @ethDreamer 

I've introduced an `AbstractSidecar` trait to support blinded blobs, similar to how the existing `AbstractExecPayload` works for blinded payloads.

This PR is not complete, but it would be great to get some feedback on the generic parts of it for now, as it would have quite a bit of impact on the implementation. 

I think it would be good to look at these types:

https://github.com/sigp/lighthouse/blob/59fb1af612dd9aed1eb3d615e2463bd03092ef16/consensus/types/src/blob_sidecar.rs#L21-L29
https://github.com/sigp/lighthouse/blob/59fb1af612dd9aed1eb3d615e2463bd03092ef16/consensus/types/src/signed_blob.rs#L32-L40
https://github.com/sigp/lighthouse/blob/59fb1af612dd9aed1eb3d615e2463bd03092ef16/common/eth2/src/types.rs#L1280-L1284
https://github.com/sigp/lighthouse/blob/59fb1af612dd9aed1eb3d615e2463bd03092ef16/common/eth2/src/types.rs#L1358-L1366

I've also maintained some aliases to existing types for readability, and it also allows me to make less changes to existing code:

https://github.com/sigp/lighthouse/blob/59fb1af612dd9aed1eb3d615e2463bd03092ef16/consensus/types/src/signed_blob.rs#L43-L46


Let me know your thoughts 🙏

**Comment by jimmygchen** [comment]
I'm going to do the remaining tasks in a separate PR to keep the PR size manageable (already quite big!):
- blob publishing (`submitBlindedBlock` endpoint updates)
- blob signing

**Comment by ethDreamer** [comment]
Nice work Jimmy!

> it would be great to get some feedback on the generic parts of it for now, as it would have quite a bit of impact on the implementation

Well, in my mind, you may have opened up a can of worms here. I think it's worth pausing and considering the best way forward for achieving polymorphism in this case. There are 2 options:

1) Traits (as you've done here)
2) Enums (via superstruct)

I found some background from [this article on this subject](https://www.mattkennedy.io/blog/rust_polymorphism/). The author tends to lead towards enums, and I suspect they would even more if they used `superstruct`. The main reason I want to pause and consider this is that I'm not sure if traits are the best choice in the case of blinded objects.

Originally @realbigsean created the [`ExecPayload`](https://github.com/sigp/lighthouse/blob/v4.2.0/consensus/types/src/payload.rs#L23) trait in order to abstract across blinded and unblinded payloads. That worked fine, but it got much more complicated when Capella hit and we had to abstract across the [`ExecutionPayload`](https://github.com/sigp/lighthouse/blob/v4.2.0/consensus/types/src/execution_payload.rs#L50) fork variants *and* the blinded/unblinded payloads at the same time. @michaelsproul eventually came up with [`AbstractExecPayload`](https://github.com/sigp/lighthouse/blob/v4.2.0/consensus/types/src/payload.rs#L77) to abstract across both of these dimensions. Doing this required him to [be quite clever](https://github.com/sigp/lighthouse/blob/v4.2.0/consensus/types/src/payload.rs#L85) and use [generic associated types](https://rust-lang.github.io/rfcs/1598-generic_associated_types.html), a feature which was (luckily) being mainlined into rust at exactly the time we were working on Capella. While we eventually got it to work, I do find that code somewhat difficult to read/understand, and I remember at the time wondering if we had painted ourselves into a corner by choosing generics for abstracting across blinded/unblinded payloads.

Now it could be the case that either implementation leads to complex code once you have to abstract across two dimensions. But the experiment hasn't been run, and this is perhaps an opportunity to consider if we want to run that experiment. I suspect if we tried the enum route then maybe some of the [enums you've created to encapsulate](https://github.com/sigp/lighthouse/blob/59fb1af612dd9aed1eb3d615e2463bd03092ef16/consensus/types/src/blob_sidecar.rs#L186) the dual behavior might no longer be necessary/be written in a cleaner way.

If we do try it out with enums, and then in some future fork there is a change to the `BlobSidecar`, then we'll get to find out which path is simpler across two dimensions. If it turns out that enums are easier to work with/cleaner, we might consider converting `ExecutionPayloads` in that direction as well.

If we decide to continue with traits for blinded sidecars, then I suggest the `AbstractSidcar` trait be renamed to just `Sidecar`, and `AbstractSidcar` be reserved for the case where a future fork changes the `BlobSidecar` so that we keep the same convention as `ExecPayload` / `AbstractExecPayload`.

I'm interested to hear if other people have thoughts on this :)

**Comment by jimmygchen** [comment]
Hey Mark, thanks a lot for the insights! 🙏 

Very useful to understand how we got here and your experience with `ExecPayload`. I agree with your points and definitely keen to hear from you and others and consider options before continuing - and I think you've raised some great points that I'd like to do some further experiment on. I've aleady done a few experiments on different approaches and I was also avoiding generics earlier because: 
- I also found it a bit difficult to understand, looking at how `Payload`s are handled
- Given we don't foresee adding fork versions for blob sidecars yet, it would be nice to keep it simple without using traits / generics

My last attempt before generics was using enums without `superstruct` (I was thinking we reserve the `superstruct`  option if we need to add the fork variants) but that didn't go so well and I ran into issues with SSZ decoding:

> Use an enum to wrap Blob & BlobRoot, and #[serde(flatten)] it so it uses different field name for each variant, but this doesn't work for SSZ decoding because Decode cannot be derived for enum_behaviour "transparent"- I'm thinking perhaps implementing Decode would work?

@sproul pointed out the above won't work because there's no signifier in the blob data that defines whether the sidecar should be full/blinded). So I scrapped that PR and started again with generics. Now your comments made me think again - maybe there's a way that could work using enums if I structure the data differently to the above. I'm keen to give it another try!

I'm still pretty new to Rust, and especially with generics & traits, so this is definitely an area worth exploring and I don't mind doing a few iterations. 

Here's some of my learnings from this current PR using `trait`s:

**Advantages:**

1. It's similar to what we already have with `Payload`, so it's a bit more comfortable and consistent, without having to touch `ExecutionPayload` - however like you said, if we find a better approach we could also convert `ExecutionPayload` 
2. It's not as complicated as `Payload` because we don't need forked versions *yet*.
3. It's more readable and concise than I thought it would be (but this is always biased coming from the author..)

**Disadvantages**
- The Rust generics doesn't work the same way as most other languages, and the type needs to be known at compile time, so we can't have a function return multiple variants or an list/map that stores different variants, which we can do with Rust enums. (we also need a separate proposal cache for blinded blobs because of this)
- I had to add `Payload` and `Sidecar` generic type params to `BlindedBeaconBlockAndBlobSidecars`, which I thought are unnecessary given the types here should always be `BlindedPayload` and `BlindedSidecar`, but this is required if want to allow `BlockContents` functions to return variants of `Payload`s and `Sidecar` - I think using enum may potentially improve this:

https://github.com/sigp/lighthouse/blob/59fb1af612dd9aed1eb3d615e2463bd03092ef16/common/eth2/src/types.rs#L1495-L1502
https://github.com/sigp/lighthouse/blob/59fb1af612dd9aed1eb3d615e2463bd03092ef16/common/eth2/src/types.rs#L1289-L1311

Agree with re-naming if we continue with the `trait` approach (definitely came across my mind.. 😅)

**Comment by ethDreamer** [comment]
> @sproul pointed out the above won't work because there's no signifier in the blob data that defines whether the sidecar should be full/blinded)

Yeah I definitely thought `Decode` might be where you would run into issues. However correct me if I'm wrong but blinded blobs are only ever used during block production right? Thus they:
1) never get stored in the DB
2) are never received from the network

If those are true then the only time you `Decode` a blob is when you receive it from gossip as a full blob. In which case you can implement a simple `Decode` that always expects a full blob.

**Comment by realbigsean** [comment]
> However correct me if I'm wrong but blinded blobs are only ever used during block production right

At some point we will probably use SSZ in the block production APIs (https://github.com/sigp/lighthouse/issues/3701) so we probably still want `Decode` for blinded blobs even if we aren't using it today. ~~Although since blinded blobs will be small maybe it's OK if it was uniquely non-SSZ but that'd be a little weird~~ Edit: I forgot blocks + blobs are combined in the API so I think we are gonna want to use SSZ for blocks + blobs there eventually

**Comment by realbigsean** [comment]
I am generally in support of exploring the enum approach because the payload types are pretty confusing. I think `Decode` generally forces us to put all the type information in the outermost type we're decoding, which is why we end up using traits or top-level enums (superstruct). I can only think of this working by making our top-level enum include blinded and non-blinded variants for each fork. But this would be a big pain because right now we do a lot per-fork logic by matching on the top-level enum. I wonder if this sort of thing could work ? `BeaconBlock::Capella(BeaconBlockCapella::Blinded(BeaconBlockCapellaBlinded {..}))` generated using `superduperstruct`?

**Comment by jimmygchen** [comment]
@ethDreamer @realbigsean 

I did a small experiment with enums using `superstruct` on this repo and documented some of my learnings on the `README.md`:
https://github.com/jimmygchen/superstruct-enum-experiment

I think the enum approach above does seem simpler to understand but potentially comes at a cost of increased code duplication (with my implementation anyway), and could get worse if we introduce a new type for EIP-6110 (in-protocol deposits)! There might be a nicer way of doing this. IMO the traits + generics approach doesn't seem too bad - at least it hides the complexity in one place 😅 . 

I don't see a clear win here between the two approaches (`enum` vs `traits`) to justify a switch. I do agree enum is generally the nicer way to do polymorphism in Rust, but this becomes less clear adding an additional dimension (block type) into the mix. The third approach with `superduperstruct` seems to require more work and could be more complicated than needed. At this point, I think it might be worth taking the 6110 changes into account before we switch to a different approach.

@michaelsproul  raised a few interesting points:
- we store finalized blocks without their execution payload by default (unless user adds the `--prune-payloads false` flag)
- the in-protocol deposit feature may require a third block type (`sparse`), although I don't understand it enough to make a good judgement based on this, but I think given there are still quite a bit unknowns there, it _might_ be worth considering to postpone making big changes until we understand the `6110` changes a bit more, and deliver Deneb first using the existing familiar approach.

More from @michaelsproul re `6110` changes:

> another thing to consider (which might give weight to the trait approach) is which abstraction fits best with making some types of blocks non-executable
>
> atm both blinded and full blocks can go through the state transitions. But with 6110 only sparse and full will actually be executable
>
> so like `fn process_block<T: ExecutableBlock>(block: T)` might be better than `fn process_block(block: SignedBeaconBlock)` where `SignedBeaconBlock` includes variants that we have to error on

In summary, I think it's definitely an experiment worth doing, but at present I don't have a better way that justifies the switch, and given `6110` might make it in the next fork after Deneb, we might want some more time to think about this.

Keen to hear your thoughts!

**Comment by sproul** [comment]
hey guys, I have been included in this conversation by accident and will unsubscribe now. Looks like I've been confused for @michaelsproul

File: consensus/types/src/payload.rs [hunk 1]
```diff
@@ -955,3 +961,84 @@ impl<T: EthSpec> From<BlindedPayload<T>> for ExecutionPayloadHeader<T> {
         }
     }
 }
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(untagged)]
+#[serde(bound = "E: EthSpec")]
+pub enum FullPayloadContents<E: EthSpec> {
```
Comment (jimmygchen) [inline consensus/types/src/payload.rs:978]: This is the response type from the builder `submitBlindedBlock` endpoint, so I think this is probably the best place for it? (same crate as `BuilderBid`)

File: consensus/types/src/payload.rs [hunk 2]
```diff
@@ -955,3 +961,84 @@ impl<T: EthSpec> From<BlindedPayload<T>> for ExecutionPayloadHeader<T> {
...
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(bound = "E: EthSpec")]
+pub struct ExecutionPayloadAndBlobs<E: EthSpec> {
+    pub execution_payload: ExecutionPayload<E>,
+    pub blobs_bundle: BlobsBundle<E>,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(bound = "E: EthSpec")]
+pub struct BlobsBundle<E: EthSpec> {
```
Comment (jimmygchen) [inline consensus/types/src/payload.rs:1049]: We can probably consolidate the `BlobsBundleV1` type with this one (and rename this to V1), and just have two BlobsBundle types:

1. `JsonBlobsBundleV1` in `execution_layer`: for API versioned response
2. `BlobsBundleV1` in `consensus/types`: for internal fork versioned types

File: beacon_node/execution_layer/src/lib.rs [hunk 3]
```diff
@@ -454,12 +455,28 @@ impl<T: EthSpec> ExecutionLayer<T> {
...
+        let maybe_blobs_bundle = maybe_json_blobs_bundle
+            .cloned()
+            .map(|blobs_bundle| BlobsBundle {
+                commitments: blobs_bundle.commitments,
+                proofs: blobs_bundle.proofs,
+                blobs: blobs_bundle.blobs,
+            });
+
+        self.inner
+            .payload_cache
```
Comment (jimmygchen) [inline beacon_node/execution_layer/src/lib.rs:522]: `PayloadCache` is an LruCache with default capacity = 10. Need to consider if this is still a good default now we're caching and intentionally keeping a copy of local EL blobs. We no longer `pop` the payload from this cache since #4239. 

https://github.com/sigp/lighthouse/blob/744b1950e54a4cf80c91e1f3a8dc26128deab062/beacon_node/execution_layer/src/payload_cache.rs#L6

File: consensus/types/src/blob_sidecar.rs [hunk 4]
```diff
@@ -1,17 +1,46 @@
...
+use tree_hash::TreeHash;
 use tree_hash_derive::TreeHash;
 
+use bls::SecretKey;
+use test_random_derive::TestRandom;
+
+use crate::test_utils::TestRandom;
+use crate::{Blob, ChainSpec, Domain, EthSpec, Fork, Hash256, SignedBlobSidecar, SignedRoot, Slot};
+
+pub enum SidecarListVariant<E: EthSpec> {
```
Comment (realbigsean) [inline consensus/types/src/blob_sidecar.rs]: We should be able to get rid of both `SidecarListVariant` and `BlobsOrBlobRoots` by relying more heavily on `AbstractExecPayload` and `Sidecar`. Adding a new associated type to represent `BlobsOrBlobRoots` should work. Or you could maybe add it as an associated type to the `Sidecar` trait. Not sure which would be simplest. And you'll probably need to add a trait for this associated type so you can work with blinded and non-blinded versions. 

I'm not sure this would make it prettier, but the blinded vs full variants of the enums should always be consistent with `AbstractExecPayload` so we can guarantee this by using generics instead of enums and I think it would make it clearer what parts of a type are generic over the same thing

Comment (jimmygchen) [reply consensus/types/src/blob_sidecar.rs]: Thanks for the review! 🙏  I tried to get rid of it and struggle a bit but I'll give it another try! Might be able to get it working now after reading your network refactoring PR ☺️

Comment (realbigsean) [reply consensus/types/src/blob_sidecar.rs]: I hacked together a version of this and it's messy and incomplete but compiling if you want to use it as inspiration https://github.com/realbigsean/lighthouse/tree/moar-generics

The naming's bad and the structure could probably be improved but it might help you with some parts

Comment (jimmygchen) [reply consensus/types/src/blob_sidecar.rs]: Nice!! That's a good cleanup actually - I attempted pretty much the same thing (adding an associated type to `Sidecar`) but ran into issues with generic types conversion!

I think these trait functions here are they key you used to resolved this! Awesome, thanks I'll get this change onto this branch!

```rust
pub trait RawBlobs<T: EthSpec>: Sync + Send {
    fn from_blob_roots(roots: BlobRoots<T>) -> Self;
    fn from_blobs(blobs: Blobs<T>) -> Self;
    fn len(&self) -> usize;
    fn is_empty(&self) -> bool;
}
```

Comment (jimmygchen) [reply consensus/types/src/blob_sidecar.rs]: I'm working through some clean ups - will push a new version tomorrow

Comment (jimmygchen) [reply consensus/types/src/blob_sidecar.rs]: I've cherry-picked your commits and cleaned up the todos, please see diffs: 97ee1f49d97c990dc21c06b742198afd2362cadc.

File: consensus/types/src/blob_sidecar.rs [hunk 5]
```diff
@@ -1,17 +1,109 @@
...
+    + Hash
+    + TreeHash
+    + TestRandom
+    + Debug
+    + SignedRoot
+    + Sync
+    + Send
+    + for<'a> arbitrary::Arbitrary<'a>
+{
+    type RawBlobs: RawBlobs<E>;
```
Comment (realbigsean) [inline consensus/types/src/blob_sidecar.rs]: I know this was my name, but I think it's confusing - it makes me think of raw bytes but it's more like "list of blobs that aren't in sidecars", which I don't currently have a good name for... what about going for `Blobs` here and renaming these type aliases:

```
pub type BlobsList<T> = VariableList<Blob<T>, <T as EthSpec>::MaxBlobsPerBlock>;
pub type BlobRootsList<T> = VariableList<Hash256, <T as EthSpec>::MaxBlobsPerBlock>;
```

File: consensus/types/src/builder_bid.rs [hunk 6]
```diff
@@ -1,21 +1,48 @@
...
+    ForkVersionDeserialize, Hash256, KzgProofs, SignedRoot, Uint256,
+};
+
 #[derive(PartialEq, Debug, Serialize, Deserialize, TreeHash, Clone)]
-#[serde(bound = "E: EthSpec, Payload: ExecPayload<E>")]
+#[serde(bound = "E: EthSpec")]
+pub struct BlindedBlobsBundle<E: EthSpec> {
+    pub commitments: KzgCommitments<E>,
+    pub proofs: KzgProofs<E>,
+    pub blob_roots: VariableList<Hash256, E::MaxBlobsPerBlock>,
```
Comment (realbigsean) [inline consensus/types/src/builder_bid.rs]: Can use the `BlobRoots` type alias here

File: consensus/types/src/builder_bid.rs [hunk 7]
```diff
@@ -37,40 +64,31 @@ pub struct SignedBuilderBid<E: EthSpec, Payload: AbstractExecPayload<E>> {
...
-        let payload_header =
-            ExecutionPayloadHeader::deserialize_by_fork::<'de, D>(helper.header, fork_name)?;
-
-        Ok(Self {
-            header: Payload::try_from(payload_header).map_err(convert_err)?,
-            value: helper.value,
-            pubkey: helper.pubkey,
-            _phantom_data: Default::default(),
+        Ok(match fork_name {
+            ForkName::Merge => Self::Merge(serde_json::from_value(value).map_err(convert_err)?),
```
Comment (realbigsean) [inline consensus/types/src/builder_bid.rs:80]: So this looks like we lose `ExecutionPayloadHeader::deserialize_by_fork` here. I think here, we are essentially calling `ExecutionPayloadHeader::deserialize` and wrapping it in whatever `BuilderBid` variant matches the fork. So this looks like it's possible to construct types where the `BuilderBid` fork doesn't match the `ExecutionPayload` fork. 

In order to make this sort of thing impossible, we could modify the `BuilderBid` to be more similar to `BeaconBlockBody`. `BeaconBlockBody` (like `BuilderBid`) is a superstruct that has a payload as a field. it uses the `ExecutionPayloadFork` structs rather than the `ExecutionPayload` enum for each of its variants. This makes sure we can't have inconsistent types (where the outer struct fork doesn't match the inner struct fork). 

I took a stab at it in this commit: d38621f4a2bbe74aee45af623e467b889fb5454a

The fact that `BuilderBid` is now a superstruct actually lets us get rid of the ugly `BlindedPayloadAsHeader` stuff which is nice

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 8]
```diff
@@ -4949,67 +4952,52 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
...
-                    blobs.len(),
-                    expected_kzg_commitments.len()
-                )));
-            }
+        let maybe_sidecar_list = match (blobs_opt, proofs_opt) {
+            (Some(blobs_or_blobs_roots), Some(proofs)) => {
+                let expected_kzg_commitments =
+                    block.body().blob_kzg_commitments().map_err(|_| {
+                        BlockProductionError::InvalidBlockVariant(
+                            "DENEB block does not contain kzg commitments".to_string(),
```
Comment (realbigsean) [inline beacon_node/beacon_chain/src/beacon_chain.rs]:
```suggestion
                            "deneb block does not contain kzg commitments".to_string(),
```

File: consensus/types/src/payload.rs (hunk 2 above)
Comment (realbigsean) [reply consensus/types/src/payload.rs:1049]: That sounds like a good idea to me! Although keeping the version without `V1` makes more sense to me, maybe we can re-introduce it if this type has to change between forks

File: consensus/types/src/payload.rs (hunk 1 above)
Comment (realbigsean) [reply consensus/types/src/payload.rs:978]: Seems good for now, maybe in the future we'll want to separate out all the builder types. Or maybe ePBS means we need similar here anyways.

**Review by realbigsean — CHANGES_REQUESTED** [review]
Great job Jimmy! Gave it a more thorough review today

File: beacon_node/execution_layer/src/lib.rs (hunk 3 above)
Comment (realbigsean) [reply beacon_node/execution_layer/src/lib.rs:522]: I think this is still alright, the max cache size would still only be something like 10MB if my math is right

Comment (jimmygchen) [reply beacon_node/execution_layer/src/lib.rs:522]: Yep that sounds about right! Thanks

File: consensus/types/src/payload.rs (hunk 2 above)
Comment (jimmygchen) [reply consensus/types/src/payload.rs:1049]: Done in 803045be8563f7551399912ed519485ba71f5116

File: consensus/types/src/blob_sidecar.rs (hunk 5 above)
Comment (jimmygchen) [reply consensus/types/src/blob_sidecar.rs]: Renaming the trait to `BlobItems` and the type aliases to the above suggestions.  Please see commit d3befb05fa1133fe92d7723f735f8bdb19c35f69.

File: consensus/types/src/builder_bid.rs (hunk 6 above)
Comment (jimmygchen) [reply consensus/types/src/builder_bid.rs]: Done in e0aba921a9e9c670ab099d4ac1376098c72b0ff4

File: consensus/types/src/builder_bid.rs (hunk 7 above)
Comment (jimmygchen) [reply consensus/types/src/builder_bid.rs:80]: Nice, good catch - that looks good! Thanks 🙏  
Updated in 3b9de468107f47b23d0bc8bee4defb1dbcdfdae4.

**Comment by jimmygchen** [comment]
Thanks for the review @realbigsean ! I've addressed your comments, please take another look when you have time. [Diffs here](https://github.com/sigp/lighthouse/compare/1882540a0dc1acf21fbb1512e49ec9d059329a9c..3b9de468107f47b23d0bc8bee4defb1dbcdfdae4).

**Comment by jimmygchen** [comment]
ah looks like I'm getting some weird test failures! On windows it's a state access violation, and I'm getting invalid memory reference on MacOS locally, and in linux the tests just seem to be weird comparisons. Will look into this tomorrow.

**Comment by realbigsean** [comment]
Ok the issue is with the changes I made to `BuidlerBid`. The type we're embedding now is `BlindedPayloadFork` rather than `ExecutionPayloadHeaderFork`. So we are failing deserialization because we have this:

```
{
  "message": {
    "header": {
       "execution_payload_header": {
         ..
       },
       ..
    },
    ..
  }
}
```

Investigating this made me realize we don't need `BuilderBid` to be generic over blinded/non-blinded at all.. it should always be blinded. I'm not sure why I originally wrote it like that. I made some updates to remove its generic here, and it fixed this deserialization issue: d00383cc35a4647b453c84135fb516b0f7fd938c

**Comment by jimmygchen** [comment]
Nice, make sense! It was an oversight on my part, didn't notice that, but the release tests look happy now 🎉
